terraform/context_test.go

```diff
@@ -117,6 +119,134 @@ func TestNewContextRequiredVersion(t *testing.T) {
 	}
 }
 
+func TestNewContext_lockedDependencies(t *testing.T) {
+	configBeepGreaterThanOne := `
+terraform {
+  required_providers {
+    aws = {
+      source  = "example.com/foo/beep"
+      version = ">= 1.0.0"
+    }
+  }
+}
+`
+	configBeepLessThanOne := `
+terraform {
+  required_providers {
+    aws = {
+      source  = "example.com/foo/beep"
+      version = "< 1.0.0"
+    }
+  }
+}
+`
+	configBuiltin := `
+terraform {
+  required_providers {
+    terraform = {
+      source = "terraform.io/builtin/terraform"
+	}
+  }
+}
+`
+	locksBeepGreaterThanOne := `
+provider "example.com/foo/beep" {
+	version     = "1.0.0"
+	constraints = ">= 1.0.0"
+	hashes = [
+		"h1:does-not-match",
+	]
+}
+`
+	beepAddr := addrs.MustParseProviderSourceString("example.com/foo/beep")
+
+	testCases := map[string]struct {
+		Config       string
+		LockFile     string
+		DevProviders []addrs.Provider
+		WantErr      string
+	}{
+		"dependencies met": {
+			Config:   configBeepGreaterThanOne,
+			LockFile: locksBeepGreaterThanOne,
+		},
+		"no locks given": {
+			Config: configBeepGreaterThanOne,
+		},
+		"builtin provider with empty locks": {
+			Config:   configBuiltin,
+			LockFile: `# This file is maintained automatically by "terraform init".`,
+		},
+		"development provider with empty locks": {
```

Comment:
Another test case to add could be where there's a managed provider and a development provider? (to ensure that the managed provider is still checked)

Reply:
Done, thanks. That's also a more realistic test case, I think.

Reply:
One more request for the path where there would be an error for the managed? That is, boop doesn't match, and the error only mentions boop and not beep. (excellent test names)